kore/src/GraphTraversal.hs

```diff
@@ -0,0 +1,319 @@
+      Stopped a [a]
+    deriving stock (Eq, Show)
+
+instance Functor TransitionResult where
+    fmap f = \case
+        StraightLine a -> StraightLine $ f a
+        Branch a as -> Branch (f a) $ NE.map f as
+        Stuck a -> Stuck $ f a
+        Final a -> Final $ f a
+        Stopped a as -> Stopped (f a) (map f as)
+
+instance Pretty a => Pretty (TransitionResult a) where
+    pretty = \case
+        StraightLine a -> single "StraightLine" a
+        Branch a as -> multi "Branch" "node" a "successors" (NE.toList as)
+        Stuck a -> single "Stuck" a
+        Final a -> single "Final" a
+        Stopped a as -> multi "Stopped" "node" a "successors" as
+      where
+        single :: Doc x -> a -> Doc x
+        single lbl a =
+            Pretty.vsep [lbl, Pretty.indent 4 $ Pretty.pretty a]
+
+        multi :: Doc x -> Doc x -> a -> Doc x -> [a] -> Doc x
+        multi lbl lbl1 a lbl2 as =
+            Pretty.vsep $
+                [ lbl
+                , Pretty.indent 2 $ "- " <> lbl1
+                , Pretty.indent 4 $ Pretty.pretty a
+                , Pretty.indent 2 $ "- " <> lbl2
+                ]
+                    <> map (Pretty.indent 4 . Pretty.pretty) as
+
+-- Graph traversal would always stop at Terminal/Cut, and _may_ stop
+-- at Branch, depending on configuration.
+
+isStuck, isFinal, isStopped, isBranch :: TransitionResult a -> Bool
+isStuck (Stuck _) = True
+isStuck _ = False
+isFinal (Final _) = True
+isFinal _ = False
+isStopped (Stopped _ _) = True
+isStopped _ = False
+isBranch (Branch _ _) = True
+isBranch _ = False
+
+extractNext :: TransitionResult a -> [a]
+extractNext = \case
+    StraightLine a -> [a]
+    Branch _ as -> NE.toList as
+    Stuck _ -> []
+    Final _ -> []
+    Stopped _ as -> as
+
+extractState :: TransitionResult a -> Maybe a
+extractState = \case
+    StraightLine _ -> Nothing
+    Branch a _ -> Just a
+    Stuck a -> Just a
+    Final a -> Just a
+    Stopped a _ -> Just a
+
+type Step prim = [prim]
+
+----------------------------------------
+transitionLeaves ::
+    forall c prim.
+    -- | Stop critera, in terms of 'TransitionResult's. The algorithm
+    -- will _always_ stop on 'Stuck' and 'Final', so [isTerminal,
+    -- isCut, isBranch] could be used here. Could simplify this to
+    -- FinalNodeType
+    [TransitionResult ([Step prim], c) -> Bool] ->
+    -- queue updating parameters,
+    -- we construct enqueue :: [a] -> Seq a -> m (Either LimitExceeded (Seq a)) from it
+    -- m is probably only there for logging purposes
+
+    -- | BreadthFirst, DepthFirst
+    GraphSearchOrder ->
+    -- | breadth limit, essentially a natural number
+    Limit Natural ->
+    -- | transition function
+    (([Step prim], c) -> Simplifier (TransitionResult ([Step prim], c))) ->
+    -- again, m is probably only for logging
+
+    -- | max-counterexamples, included in the internal logic
+    Natural ->
+    -- | initial node
+    ([Step prim], c) ->
```

Comment:
What is the `[Step prim]`? I assume `c` is the configuration state.

Reply:
`prim` is the "mode" of transition (`Simplify`, `Apply*`, `CheckImplications`), a `Step` is one series of transitions with different modes that are executed together. Will change `c` to `config`.